### DISCUSSION TITLE
Cache miss even though in initial SSR data

### DISCUSSION THREAD

**Discussion by tjk** [urql-graphql/urql#2024] [Q&A]
![image (3)](https://user-images.githubusercontent.com/109224/136892170-80851a0f-20ac-4268-9dc1-ca52f7da2185.png)

Any ideas why this might happen? I can share code if needed... (originally mentioned in https://github.com/FormidableLabs/urql/issues/1512#issuecomment-940524850)

Using:

```js
[
dedupExchange,
cacheExchange,
ssrExchange,
persistedFetchExchange,
fetchExchange,
]
```

**Comment by JoviDeCroock** [comment]
Would be easier with some way to verify the ssr-cache vs the oepration-keys you're asking for 😅 all we can tell form this screenshot is that 2 misses happened, what their `operation.key` was and what they were querying in relation to the `hits` might be important context.

**Reply by tjk** [reply]
Do you see the logged initial data at the bottom?

**Reply by tjk** [reply]
(The op key for at least the first cache miss is in the SSR initial data)

**Reply by JoviDeCroock** [reply]
Yea but what keys are being requested? We can see two misses and two hits, which is expected given that there are only two keys in ssr-data

**Reply by tjk** [reply]
I do use `cache-and-network` as the default request policy... but I don't expect promises to block at the beginning (should just resolve with stale data immediately) or for the debug messages to say "cache miss".

**Reply by tjk** [reply]
```
cacheMiss 1651902451
cacheHit 1651902451
cacheMiss 1042488748
cacheHit 1042488748
fetchRequest 1651902451
fetchRequest 1042488748
fetchSuccess 1042488748
fetchSuccess 1651902451
cacheHit 1042488748
> Object.keys(__urql__)
(2) ['1042488748', '1651902451']
```

**Reply by tjk** [reply]
Maybe this is the expected behavior with `cache-and-network`? I need to verify but it looks like the promises do not resolve immediately though which I do not expect.

**Reply by JoviDeCroock** [reply]
As I said, it's a bit hard currently to give you an answer as it's hard to see what exactly goes on here. Generally however I can say that `cache-and-network` should not go into effect unless the cache hits inside of the `cacheExchange` so this might be due to two queries requesting the same resource, the ssr-exchange won't revalidate data unless you ask it to https://github.com/FormidableLabs/urql/blob/main/packages/core/src/exchanges/ssr.ts#L142 and the ssr-exchange evicts data after the first time it gets requested because it knows that by then it will be written to cache. 

Those are the things I'm thinking about when tackling this issue, so data will be set to stale if `staleWhileRevalidate` is true or when a background-refetch happens due to the second invocation going to an update mechanism.

Just throwing this out there but if always refetching isn't ideal we have solutions for selective upgrading https://github.com/FormidableLabs/urql/tree/main/exchanges/request-policy

EDIT: afaik there are no promises in this code apart from the `fetchExchange` all the other communication is stream-based, you might be referring to `stale` being `true` as a Promise maybe (?) but yes this does look like the expected behavior outlined in the first paragraph

**Reply by tjk** [reply]
Hm, using `requestPolicyExchange` (and the default `requestPolicy`), on the client I see:

```
cacheMiss 1651902451
cacheMiss 1042488748
cacheHit 1651902451
```

which is a bit unexpected (not sure why miss+hit don't match) and I sort of would have liked the refresh to happen but this behavior might be an improvement since shouldn't need to refresh when hydrating.

**Reply by tjk** [reply]
~~After hydration, I load a page, leave, and return (before the configured 1000ms ttl) and see this which doesn't make any sense~~ Ignore this -- I had one second TTL instead of a minute...

```
cacheHit 3011224006
cacheHit 865969309
cacheHit 3417702101
fetchRequest 3011224006
fetchRequest 865969309
fetchRequest 3417702101
```

**Reply by tjk** [reply]
Anyway, thanks for your help -- I might do some logging within `urql` to try to understand this behavior.
